### PR TITLE
fix issue that no available connections cause by concurrency request limit bug

### LINKED ISSUE #1225: no available connections cause by concurrency request limit bug
# Bug Report

#1098 introduced concurrency request limit.

https://github.com/tikv/client-go/blob/03bbadb2777ec8ab003123f24bc828dc48c03c31/internal/client/client_batch.go#L407

https://github.com/tikv/client-go/blob/03bbadb2777ec8ab003123f24bc828dc48c03c31/internal/client/client_batch.go#L569-L571

Following is the general situation about the upper code:

```go
hasHighPriorityTask = false  // most request don't have high priority.
c.maxConcurrencyRequestLimit = math.MaxInt64 // by default
c.sent >= 0
```

In general, the above code works fine. But it will have a bug with the data race in #1222 : 

- the data race will cause the `c.sent` to be less than 0
- then it will cause `c.available()` computation overflow
- then `c.available() > 0` will be false.
- then have `no available connections` issue, and the request could not be sent, resulting in a QPS drop to 0.  See issue in https://github.com/pingcap/tidb/issues/51728.

### PR DESCRIPTION
**Pull request by crazycs520**
Close #1225.

Data race issue #1222 should be fix in PR #1219.

### CODE REVIEW

File: internal/client/client_batch.go [hunk 1]
```diff
@@ -567,7 +567,14 @@ func (c *batchCommandsClient) isStopped() bool {
 }
 
 func (c *batchCommandsClient) available() int64 {
-	return c.maxConcurrencyRequestLimit.Load() - c.sent.Load()
+	limit := c.maxConcurrencyRequestLimit.Load()
+	sent := c.sent.Load()
+	// just in case, `sent` may less than 0, see https://github.com/tikv/client-go/issues/1225
```
Comment (cfzjywxk) [inline internal/client/client_batch.go:572]:
```suggestion
	//  The `sent` could be less than 0, see https://github.com/tikv/client-go/issues/1225 for details.
```

File: internal/client/client_batch.go [hunk 2]
```diff
@@ -567,7 +567,14 @@ func (c *batchCommandsClient) isStopped() bool {
...
+	// then maxInt64 - (-1) will overflow, which is unexpected.
+	if sent > 0 {
+		return limit - sent
```
Comment (cfzjywxk) [inline internal/client/client_batch.go:575]: Is it acceptable to return a negative result?

**Comment by ekexium** [comment]
I didn't get it very well. Why are we fixing it by checking `sent > 0`? Seems to me it should be fixed by making sure `sent` is always non-negative?

**Comment by crazycs520** [comment]
> I didn't get it very well. Why are we fixing it by checking `sent > 0`? Seems to me it should be fixed by making sure `sent` is always non-negative?

`sent` may be negative is cause by data race in issue #1222 , and I also have PR #1219 to do this fix, but I need to to more test to avoid regression as mentioned in https://github.com/tikv/client-go/pull/1219#discussion_r1523342977 , so I try to fix this bug without fix data race in this PR, since  code freezes.

File: internal/client/client_batch.go (hunk 2 above)
Comment (crazycs520) [reply internal/client/client_batch.go:574]: It's ok here, negative result means not available.

Comment (crazycs520) [reply internal/client/client_batch.go:574]: Nice cache, it's better not return negative result here.

**Review by ekexium — APPROVED** [review]

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**